### PR TITLE
REF: prepare dataframe info for series info

### PR DESCRIPTION
**Pull request by ivanovmg**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Precursor for #37320

Refactor ``pandas/io/formats/info.py`` to simplify further introduction of ``Series.info``.
Basically what is done here:
 - Move docstring to ``BaseInfo`` from ``DataFrameInfo`` - supposedly will be shared with ``SeriesInfo``
 - Move ``col_count``, ``ids`` from ``BaseInfo`` to ``DataFrameInfo`` (will not be relevant to ``SeriesInfo``)
 - Move ``memory_usage_bytes`` to ``DataFrameInfo`` (will have different implementation for ``SeriesInfo``)
 - Extract ``InfoPrinterAbstract``. Concrete classes will differ by the ``_create_table_builder``
 - Move common methods and properties from ``DataFrameTableBuilder`` to ``TableBuilderAbstract``
 - Extract ``TableBuilderVerboseMixin`` from ``DataFrameTableBuilderVerbose`` as the generator functions there will be used for verbose series info builders

I suggest moving the substitution decorator for the docstring from ``pandas/core/frame.py`` to ``pandas/io/formats/info.py`` later on in a separate PR.

### CODE REVIEW

File: pandas/io/formats/info.py [hunk 1]
```diff
@@ -87,26 +96,26 @@ class BaseInfo(ABC):
...
-        self.memory_usage = _initialize_memory_usage(memory_usage)
+    data: FrameOrSeriesUnion
+    memory_usage: Union[bool, str]
 
     @property
-    @abstractmethod
-    def ids(self) -> Index:
-        """Column names or index names."""
+    def dtypes(self) -> Iterable[Dtype]:
+        """Dtypes.
```
Comment (jbrockmendel) [inline pandas/io/formats/info.py:112]: newline after triple-quote

if there's nothing useful to write, leave it blank

Comment (ivanovmg) [reply pandas/io/formats/info.py]: Changed that.

File: pandas/io/formats/info.py [hunk 2]
```diff
@@ -83,30 +92,32 @@ def _initialize_memory_usage(
...
 
 class BaseInfo(ABC):
-    """Base class for DataFrameInfo and SeriesInfo.
+    """
+    Base class for DataFrameInfo and SeriesInfo.
 
     Parameters
     ----------
-    data : FrameOrSeries
+    data : FrameOrSeriesUnion
```
Comment (jreback) [inline pandas/io/formats/info.py:100]: @simonjayhawkins is this right here?

**Review by jreback — CHANGES_REQUESTED** [review]
looks fine.

File: pandas/io/formats/info.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/io/formats/info.py:100]: The variable type annotation for `data` is correct if this doesn't need to be a Generic class.

FrameOrSeries is a typevar and using a typevar for a variable type annotation gives `error: Type variable "pandas._typing.FrameOrSeries" is unbound  [valid-type]`

However, for the docstring, I think it is still preferred not to use Type notation, even for internal docstrings and follow the documentation contributing guide.

File: pandas/core/frame.py [hunk 3]
```diff
@@ -159,7 +159,7 @@
 
 from pandas.io.common import get_handle
 from pandas.io.formats import console, format as fmt
-from pandas.io.formats.info import DataFrameInfo
+from pandas.io.formats.info import BaseInfo, DataFrameInfo
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:162]: do we need to add the import of BaseInfo just for the docstring? could the docstring from BaseInfo.render be added to DataFrameInfo.render (could be a follow-on)

Comment (ivanovmg) [reply pandas/core/frame.py:162]: Yes, I plan to fix it in the follow-up.
I struggled with ``doc`` decorator, so I could not figure out how to make it work properly so far.
I will create another PR on this and would be very grateful if you would suggest there how to resolve the outstanding issues.

File: pandas/io/formats/info.py (hunk 2 above)
Comment (ivanovmg) [reply pandas/io/formats/info.py]: Changed that.

**Review by simonjayhawkins — APPROVED** [review]
Thanks @ivanovmg. lgtm. one comment (can be done in the follow-up)

File: pandas/io/formats/info.py [hunk 4]
```diff
@@ -83,30 +92,32 @@ def _initialize_memory_usage(
...
-        self.data = data
-        self.memory_usage = _initialize_memory_usage(memory_usage)
+    data: FrameOrSeriesUnion
+    memory_usage: Union[bool, str]
 
     @property
-    @abstractmethod
-    def ids(self) -> Index:
-        """Column names or index names."""
+    def dtypes(self) -> Iterable[Dtype]:
```
[the remark this answers is not in the document]
Comment (ivanovmg) [reply pandas/io/formats/info.py:112]: Yes, I guess ``abstractmethod`` should be there.

**Comment by ivanovmg** [comment]
Got some irrelevant error in CI Check.

```
mypy --version
mypy 0.782
Performing static analysis using mypy
pandas/core/indexes/datetimelike.py:776: error: Argument 1 to "_simple_new" of "DatetimeIndexOpsMixin" has incompatible type "Union[ExtensionArray, Any]"; expected "Union[DatetimeArray, TimedeltaArray, PeriodArray]"  [arg-type]
Found 1 error in 1 file (checked 1119 source files)
```

**Comment by simonjayhawkins** [comment]
> Got some irrelevant error in CI Check.

fixed on master

File: pandas/io/formats/info.py [hunk 5]
```diff
@@ -83,30 +92,33 @@ def _initialize_memory_usage(
...
     """
 
-    def __init__(
-        self,
-        data: FrameOrSeries,
-        memory_usage: Optional[Union[bool, str]] = None,
-    ):
-        self.data = data
-        self.memory_usage = _initialize_memory_usage(memory_usage)
+    data: FrameOrSeriesUnion
```
Comment (jreback) [inline pandas/io/formats/info.py:108]: are we using FrameOrSeries? cc @simonjayhawkins

Comment (simonjayhawkins) [reply pandas/io/formats/info.py:108]: see https://github.com/pandas-dev/pandas/pull/37868#discussion_r525126264

File: pandas/core/frame.py [hunk 6]
```diff
@@ -2523,16 +2523,25 @@ def to_html(
     @Substitution(
         klass="DataFrame",
         type_sub=" and columns",
-        max_cols_sub=(
-            """max_cols : int, optional
+        max_cols_sub=dedent(
```
Comment (jreback) [inline pandas/core/frame.py:2526]: can you do a followup where you switch to using `@doc` before the Series.info change

**Review by jreback — APPROVED** [review]

**Pull request merged**